web-console/src/lib/services/pipelineManagerQuery.ts

```diff
@@ -63,62 +66,117 @@ const consolidatePipelineStatus = (newPipeline: Pipeline, oldPipeline: Pipeline
+}
+
+export const PublicPipelineManagerQuery = mkQuery({
+  getAuthConfig: () =>
+    AuthenticationService.getAuthenticationConfig().then((c: AuthProvider) =>
+      match(c)
+        .with({ AwsCognito: P.select() }, config => ({
+          AwsCognito: {
+            ...config,
+            // AWS Cognito's default grant type is `code`
+            grantType: assertUnion(
+              ['token', 'code'] as const,
+              /response_type=(token|code)/.exec(config.login_url)?.[1] ?? 'code'
+            )
+          }
+        }))
+        .with({ GoogleIdentity: P.select() }, (_, value) => value)
+        .with({} as any, () => 'NoAuth' as const) // Config may be an empty record
+        .exhaustive()
+    )
+})
+
+const PipelineManagerApi = {
+  programs: () => ProgramsService.getPrograms(),
+  programCode: (programName: string) => ProgramsService.getProgram(programName, true),
+  programStatus: (programName: string) => ProgramsService.getProgram(programName, false),
+  pipelines: () =>
+    PipelinesService.listPipelines().then(ps =>
+      ps.map(p => ({
+        ...p,
+        state: { ...p.state, current_status: toClientPipelineStatus(p.state.current_status) }
+      }))
     ),
-    pipelineConfig: PipelinesService.getPipelineConfig,
-    pipelineStats: PipelinesService.pipelineStats,
-    pipelineLastRevision: PipelinesService.pipelineDeployed,
-    pipelineValidate: PipelinesService.pipelineValidate,
-    connectors: () => ConnectorsService.listConnectors(),
-    connectorStatus: ConnectorsService.getConnector,
-    getAuthConfig: AuthenticationService.getAuthenticationConfig,
-    listApiKeys: ApiKeysService.listApiKeys,
-    getApiKey: ApiKeysService.getApiKey,
-    listServices: ServicesService.listServices,
-    getService: ServicesService.getService,
-    newService: ServicesService.newService
-  })
-)
+  pipelineStatus: compose(PipelinesService.getPipeline, p =>
+    p.then(p => ({ ...p, state: { ...p.state, current_status: toClientPipelineStatus(p.state.current_status) } }))
+  ),
+  pipelineConfig: PipelinesService.getPipelineConfig,
+  pipelineStats: PipelinesService.pipelineStats,
+  pipelineLastRevision: PipelinesService.pipelineDeployed,
+  pipelineValidate: PipelinesService.pipelineValidate,
+  connectors: () => ConnectorsService.listConnectors(),
+  connectorStatus: ConnectorsService.getConnector,
+  listApiKeys: ApiKeysService.listApiKeys,
+  getApiKey: ApiKeysService.getApiKey,
+  listServices: ServicesService.listServices,
+  getService: ServicesService.getService,
+  newService: ServicesService.newService
+}
+
+export const makePipelineManagerQuery = ({
+  onError = defaultPipelineManagerQueryOptions.onError
+}: PipelineManagerQueryOptions) =>
+  (({ pipelines, pipelineStatus, ...queries }) => ({
+    ...queries,
+    pipelines: () => ({
+      ...pipelines(),
+      // Avoid displaying PROVISIONING status when local status is more detailed
+      structuralSharing<T>(oldData: T | undefined, newData: T) {
```

Comment:
couldn't find this structuralSharing function with a quick search.
What does it do?

Reply:
It is an optional argument of useQuery hook that enables reconciling a new query result with a previous one. Here I use it to keep a locally set pipeline status when server returns generic "provisioning" status